File: snc/android/app/src/main/kotlin/com/shortnerdcat/snc/NavlinkAuth.kt
// The Tunnel Cat Project
// Copyright (C) NavLink, 2026
// Лицензировано под лицензией Apache 2.0

package com.shortnerdcat.snc

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.HttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Minimal in-memory cookie jar: navlink.net sets one session cookie on login
 * that must be replayed on the very next call ([freeKey]). No persistence
 * across process restarts is needed for that, so a plain per-host list is
 * enough — avoids pulling in the separate okhttp-urlconnection artifact just
 * for JavaNetCookieJar.
 */
private class SimpleCookieJar : CookieJar {
    private val store = mutableMapOf<String, List<Cookie>>()

    override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
        store[url.host] = cookies
    }

    override fun loadForRequest(url: HttpUrl): List<Cookie> = store[url.host] ?: emptyList()
}

/**
 * Direct (non-tunneled) HTTPS client for navlink.net's existing account-login
 * and free-key-issuance endpoints — the "no key yet" login path.
 *
 * Deliberately its own [OkHttpClient], independent of anything SNC tunnel
 * related: a device with no key yet has no VPN/tunnel running (SNCVpnService
 * is only started once a key exists), so this client is structurally
 * guaranteed to go straight over the device's normal network stack. Never
 * wire this through a tunnel-aware client, even if one becomes reachable
 * later in the process.
 */
object NavlinkAuth {
    private const val BASE = "https://navlink.net"
    private val JSON = "application/json".toMediaType()

    private val client = OkHttpClient.Builder()
        .cookieJar(SimpleCookieJar())
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()

    class NavlinkException(val statusCode: Int, message: String) : Exception(message)

    /**
     * Reports whether navlink.net is reachable directly, right now. Any HTTP
     * response (regardless of status code) counts as reachable — only a
     * network/TLS-level failure or timeout counts as unreachable.
     */
    suspend fun probe(): Boolean = withContext(Dispatchers.IO) {
        try {
            val probeClient = client.newBuilder()
                .connectTimeout(4, TimeUnit.SECONDS)
                .readTimeout(4, TimeUnit.SECONDS)
                .build()
            val req = Request.Builder().url("$BASE/").get().build()
            probeClient.newCall(req).execute().use { true }
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Authenticates an existing navlink.net account. On success the session
     * cookie is retained by this object's cookie jar for a subsequent
     * [freeKey] call. Throws [NavlinkException] on a well-formed error
     * response (e.g. wrong password), or a plain IOException on network
     * failure.
     */
    suspend fun login(email: String, password: String) = withContext(Dispatchers.IO) {
        val body = JSONObject().put("email", email).put("password", password)
            .toString().toRequestBody(JSON)
        val req = Request.Builder().url("$BASE/api/account/login").post(body).build()
        client.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) {
                throw NavlinkException(resp.code, readErrorMessage(resp.body?.string()))
            }
        }
    }

    data class IssuedKey(val key: String, val keyId: String, val clientId: String)

    /** Issues a fresh key for the account authenticated by the preceding [login] call. */
    suspend fun freeKey(): IssuedKey = withContext(Dispatchers.IO) {
        val req = Request.Builder().url("$BASE/api/key/free").post(ByteArray(0).toRequestBody(null)).build()
        client.newCall(req).execute().use { resp ->
            val text = resp.body?.string() ?: ""
            if (!resp.isSuccessful) {
                throw NavlinkException(resp.code, readErrorMessage(text))
            }
            val json = JSONObject(text)
            IssuedKey(
                key = json.getString("key"),
                keyId = json.optString("key_id"),
                clientId = json.optString("client_id"),
            )
        }
    }

    private fun readErrorMessage(body: String?): String {
        if (body.isNullOrEmpty()) return "request failed"
        return try {
            JSONObject(body).optString("error", body)
        } catch (e: Exception) {
            body
        }
    }
}

File: snc/ios/SNCApp/NavlinkAuth.swift
// The Tunnel Cat Project
// Copyright (C) NavLink, 2026
// Лицензировано под лицензией Apache 2.0

import Foundation

/// Direct (non-tunneled) HTTPS client for navlink.net's existing account-login
/// and free-key-issuance endpoints — the "no key yet" login path.
///
/// Uses `URLSession(configuration: .default)`, deliberately independent of
/// anything tunnel-related: a device with no key yet has no VPN tunnel
/// running (VPNManager only connects once a key exists), so this session is
/// structurally guaranteed to go straight over the device's normal network
/// stack. The default configuration's `HTTPCookieStorage` persists the
/// session cookie set by `login` for the subsequent `freeKey` call within
/// the same process. Never wire this through a tunnel-aware session, even if
/// one becomes reachable later in the process.
enum NavlinkAuth {

    private static let baseURL = URL(string: "https://navlink.net")!

    private static let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 10
        return URLSession(configuration: config)
    }()

    struct NavlinkError: Error, LocalizedError {
        let statusCode: Int
        let message: String
        var errorDescription: String? { message }
    }

    /// Reports whether navlink.net is reachable directly, right now. Any HTTP
    /// response (regardless of status code) counts as reachable — only a
    /// network/TLS-level failure or timeout counts as unreachable.
    static func probe(completion: @escaping (Bool) -> Void) {
        var req = URLRequest(url: baseURL)
        req.httpMethod = "GET"
        req.timeoutInterval = 4
        let probeSession: URLSession = {
            let config = URLSessionConfiguration.default
            config.timeoutIntervalForRequest = 4
            return URLSession(configuration: config)
        }()
        probeSession.dataTask(with: req) { _, response, error in
            completion(error == nil && response != nil)
        }.resume()
    }

    /// Authenticates an existing navlink.net account. On success the session
    /// cookie is retained by `session`'s cookie storage for a subsequent
    /// `freeKey` call.
    static func login(email: String, password: String, completion: @escaping (Error?) -> Void) {
        var req = URLRequest(url: baseURL.appendingPathComponent("api/account/login"))
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try? JSONSerialization.data(withJSONObject: ["email": email, "password": password])

        session.dataTask(with: req) { data, response, error in
            if let error {
                completion(error)
                return
            }
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            if status != 200 {
                completion(NavlinkError(statusCode: status, message: Self.errorMessage(from: data)))
                return
            }
            completion(nil)
        }.resume()
    }

    struct IssuedKey {
        let key: String
        let keyID: String
        let clientID: String
    }

    /// Issues a fresh key for the account authenticated by the preceding
    /// `login` call, using the same session (and thus the same cookie).
    static func freeKey(completion: @escaping (Result<IssuedKey, Error>) -> Void) {
        var req = URLRequest(url: baseURL.appendingPathComponent("api/key/free"))
        req.httpMethod = "POST"

        session.dataTask(with: req) { data, response, error in
            if let error {
                completion(.failure(error))
                return
            }
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            guard status == 200, let data else {
                completion(.failure(NavlinkError(statusCode: status, message: Self.errorMessage(from: data))))
                return
            }
            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let key = json["key"] as? String else {
                completion(.failure(NavlinkError(statusCode: status, message: "malformed response")))
                return
            }
            let issued = IssuedKey(
                key: key,
                keyID: json["key_id"] as? String ?? "",
                clientID: json["client_id"] as? String ?? "")
            completion(.success(issued))
        }.resume()
    }

    private static func errorMessage(from data: Data?) -> String {
        guard let data, !data.isEmpty else { return "request failed" }
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let msg = json["error"] as? String {
            return msg
        }
        return String(data: data, encoding: .utf8) ?? "request failed"
    }
}
